internal/validation/gateway.go

```diff
@@ -0,0 +1,115 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO [danehans]: Refactor to use upstream validation pkg.
+
+package validation
+
+import (
+	"context"
+	"fmt"
+
+	retryable "github.com/projectcontour/contour/internal/retryableerror"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gateway_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+// Gateway returns an error if gw is an invalid Gateway.
+func Gateway(ctx context.Context, cli client.Client, gw *gateway_v1alpha1.Gateway) error {
```

Comment:
We need to think about where the status is going to be set for Gateways. There is some up-front validation that can be applied, but what parts do routes have an affect on the status of the gateway? 

Additionally, if we don't pass the configured gateway off to the gateway_processor, then all the other routes will be invalidated and taken out of Envoy config. That might be ok since if the gateway is invalid, then the routes probably should't get traffic, but want to understand their relationship.

Reply:
@stevesloka I'm going to consider this comment resolved since the scope of this PR was updated to focus only on the gatewayclass controller.